Change topLevel check to topActive check

kaosko opened the pull request:
- Commit changes from
https://github.com/satago/tapestry-jpa-transactions/pull/1
- add a simple testing framework that allows testing JPA down the
database level and few tests to cover nested transactions. Need better
tests for covering the commit hooks

dmitrygusev commented on build.gradle:
```diff
@@ -38,12 +38,20 @@ repositories {
     compile 'org.slf4j:slf4j-api:1.7.12'
 
-    def tapestry_version='5.3.7'
+    def tapestry_version='5.5.0-alpha-1'
```
Is it really necessary do depend on the latest 5.5 release?
I don't think we use any 5.5 features unless I missed something, in which case it should be binary compatible with 5.3 branch, right?

pedrocborges2 replied:
@kaosko Is the code going to be `merged` into the tapestry5 core or any other existing module or is it going to be available as a standalone module?

kaosko replied:
No, obviously not needed to depend on 5.5 but that's where I'm going to merge it in. I'm planning on merging the code directly into tapestry-jpa module for the 5.5 branch (currently in master).

dmitrygusev commented on build.gradle:
```diff
@@ -38,12 +38,20 @@ repositories {
+
+		testRuntime 'org.eclipse.persistence:eclipselink:2.5.1'
+		testRuntime 'com.h2database:h2:1.3.175'
+
```
can we use spaces instead of tabs for consistency?

kaosko replied:
Yes. Tapestry uses spaces. You seemed to have mixed formatting, so I didn't worry about formatting at the moment. All code needs to be reformatted accoding to Tapestry formatting conventions when merged.

dmitrygusev commented on src/test/java/net/satago/tapestry5/jpa/test/JpaTestModule.java:
```diff
@@ -0,0 +1,83 @@
+			}
+		};
+		cfg.add("jpatest", configurer);
+	}
```
By the way, in our projects we use Hibernate as JPA provider.
Can we add tests for Hibernate JPA too?

I think we can do it like this:
1) move `configurePersistenceUnit` to another module, say `EclipseLinkJpaTestModule`
2) create similar configuration for `HibernateJpaTestModule`
3) convert `JpaTest` to some sort of parameterized test, and create different configurations of registry depending on parameter (EclipseLink or Hibernate) to run full test suite on both implementations.

What do you think?

kaosko replied:
Feel free to do it. I wasn't targeting specific JPA implementations but the functionality in this module.

dmitrygusev commented on src/test/java/net/satago/tapestry5/jpa/test/NestedServiceImpl.java:
```diff
@@ -0,0 +1,23 @@
+		this.em = em;
+	}
+
+	@CommitAfter
```
I don't think it has any effect if you put `@CommitAfter` annotation here, because you already have one on the interface.
If I remember correct Tapestry can copy annotations from implementation classes to interfaces since T5.4, but then you don't need the one on the interface?

Same for the `TopLevelService/Impl`, especially for `createThingOneThenTwo` that only has `@CommitAfter` on implementation.

kaosko replied:
Correct, you don't need it in the interface.

dmitrygusev commented on src/main/java/net/satago/tapestry5/jpa/TransactionalUnit.java:
```diff
@@ -33,7 +33,7 @@
 
```
We've been discussing this fix offline with @pedrocborges2 long time ago, but never returned to it after that.

The problem we're trying to solve here is that if you try to create a new `transactionalUnit` from `afterCommit` callback there will still be previous (already committed) top level `transactionalUnit` in the `currentUnit` stack. And as a result of that this newly created unit will never be committed.

So instead of playing with the state of current transaction as in `transaction.isActive()` (which, by the way, _may_ be changed outside of transactionalUnits API via JPA interfaces) we could do `currentUnit.get().pop()` before invoking callbacks.

This should work fine for `afterCallback`, but the question is still open for what we should do if user is trying to run another transactional unit from `beforeCommit` callback. I don't have any good answer for this right now. This is an edge case and we don't do this in our applications, so we could either document this better somewhere and/or forbid creating new nested units from the callback(s) by throwing an exception.

What do you think?

kaosko replied:
Yes, I like the idea of popping them off the stack once done. A transactional unit before commit is an interesting edge case but there's no pressure to resolve all of it right now. Related to your earlier comment about T5 version, it probably makes the most sense to offer this module for pre T5.5 versions. It probably requires some amount of duplicate effort on merging any fixes but likely not a major issue.

dmitrygusev commented on src/test/java/net/satago/tapestry5/jpa/test/TopLevelServiceImpl.java:
```diff
@@ -0,0 +1,53 @@
+				em.persist(thingOne);
+
+			}
+		});
```
I have a feeling that we don't have any nested units here, because there's no `@CommitAfter` on interface method, will try to check this tomorrow. I haven't tried this "annotations on services implementations" feature yet.

And another feeling I have by looking at implementation of `TransactionalUnit` is that there may be a bug, that `afterCommit` callbacks won't be executed for nested transactions, only for top-level one.

kaosko replied:
Yeah I wanted to verify all of those but my primary focus was laying out some some simple structure for testing first.

kaosko commented:
I made a new test case for EntityListeners and looks like transactional units in @PostUpdate are not committed. Do you want to keep following in-progress work? If so, I can keep making intermediate commits. Alternatively, if you prefer to preserve the state of this module as is, you could just kill the pull request and I can instead bring what I have right now to T5.5 and work on it there.

dmitrygusev commented:
Intermediate commits are fine, let's continue with this PR.

kaosko commented:
Ok I see what you are saying about the committed but still active transaction. Creating a transactional unit in @PostUpdate has exactly the same problem and I agree that using the transaction.isActive() and the way it's done in the pull request #1 is just not going to work. I'll have to experiment with it some more. Given that this is the crux of the issue, it makes sense to continue here instead of trying to rush the merge into T5.

kaosko commented:
There are several issues with the codebase. The beforecommit and aftercommit hooks are only fired for the transactionalunit that is at the top of the stack. The fix for beforecommit should be simple and aftercommits should be handled in a separate transaction. This is more complicated for JPA @Post listeners that create transactional units since the events are fired during the commit operation while transaction is still active. Even still, transactions from multiple entityManagers are using the same stack which won't work. Since JPA doesn't support it natively, it would be useful to be able to trigger cascading sequential commit but it requires complete refactoring of how the transactional units are currently handled.

kaosko commented:
The latest push represents a semantically correct or at least consistent refactoring of the concept, with several tests backing it up. I'm going to go with a version of it for T5.4 mainline. It would certainly be useful to have more eyes on testing issues with thread safety and multiple persistence contexts in real world platforms but I fully understand that for you, it may be better to kill this pull request and let your module stand as a standalone offering for pre-T5.4, given that the implementation has drastically changed (but you should at least warn the users of the shortcomings, e.g. meant for a single persistence unit only and that the hooks are not reliably being fired).

dmitrygusev commented:
Hey Kalle,

Regarding different `entityManagers` in one transaction stack -- that was an design decision that it won't support different persistence units in one context, because otherwise it might give you a false guarantees of consistency. There were no intention to implement any support for distributed transactions. What's missing from original implementation is an assertion that throws a `RuntimeException` in case a user would try to mix different PUs in one stack -- I'll fix that in current implementation.

Note that you can still choose which PU to use in a transactional unit, it was missing from the `TransactionalUnitBuilder` though, assuming it would build a unit for default PU. That should be an easy fix too.

As for the rewrite, there are few features missing comparing to original implementation:
1. `invokeAfterCommit` & `invokeBeforeCommit` are now instance methods, which requires that classes participating in a transaction need to have a reference to an instance of `EntityTransactionManager`, meaning that you need to pass/`@Inject` it to the participating class somehow. I've checked our use cases, it's not a problem to us -- all our participants are tapestry services, but still could be a slightly limitation to some cases.
2. `invokeAfterCommit` & `invokeBeforeCommit` methods don't check that there's an active enclosing transactional unit when you register the callbacks. Meaning that newly registered afterCommit callback will never be invoked, because there's no active transaction that could trigger it, or it may be invoked as a part of another (next) transactional stack -- which could be even more misleading. This feature saved as a few times, allowing to catch some bugs early.
3. There's no more equivalent of `TransactionalUnitBuilder` that allowed constructing an instance of transactional unit with attached before/after callbacks outside of the transactional unit, decoupling a logic of transactional unit from callbacks, i.e.:

``` java
    private <T> void workInTransaction(
        final List<T> entities,
        final Worker<T> worker,
        final Runnable beforeCommit)
    {
        transactionalUnits
                .prepareRun(() -> entities.forEach(worker::work))
                .beforeCommit(beforeCommit)
                .runInTransaction();
    }
```

I couldn't find a way how you could do it with this rewrite.

Few notes/questions:
- There's no more stack trace in the nested transaction warning which would make it nearly impossible to see what caused it.
- Why is `afterCommit` callback can now interrupt the flow of other callbacks? Is it for cascaded commits?

kaosko commented:
Thanks, good points.
1) I think it's more honest that it's a service. Static operations can not be easily overridden later.
2) I was thinking that it's more flexible to be able to add the callbacks for whatever transaction happens next, but agree, perhaps its safer to not allow registering them at all if a transaction isn't open.
3) Yeah, so regarding before/after callbacks in transactional units, I couldn't figure out how to consistently execute a tree model (each unit having its all callbacks that can spawn new transactions) given that JPA doesn't allow actual nested transactions. Once I removed the callbacks from the units, there was no convincing reason for keeping the transactional unit either. Now, I'm fond of the build pattern but I didn't see a pressing need for it here. I considered chaining, and while that's not usually the style for service operations, I could change the signature to allow `entityTransactionManager.invokeBeforeTransaction(invokable).runInTransaction(runnable)`

Makes sense to add the warning about nested transactions back. Yes, I added the afterCommit callback's return value for symmetry to make it possible to explicitly stop a sequential transaction flow - though I'm not sure that the semantics are obvious to the user.

dmitrygusev commented on src/test/java/net/satago/tapestry5/jpa/test/CommitCounter.java:
```diff
@@ -0,0 +1,54 @@
+                versionedThing.setId(1);
+                versionedThing.setLastTouched(new Date());
+
+                entityManager.merge(versionedThing);
```
FYI

I'm trying to back port these tests to master and this bit doesn't work (at least with Hibernate) throwing a ConcurrentModificationException from Hibernate internals. Here's what I've found:

> To avoid conflicts with the original database operation that fires the entity lifecycle event (which is still in progress) callback methods should not call EntityMan­ager or Query methods and should not access any other entity objects.
> 
> If a callback method throws an exception within an active transaction, the transaction is marked for rollback and no more callback methods are invoked for that operation.

https://www.objectdb.com/java/jpa/persistence/event#Implementation_Restrictions_

This is exactly what's happening here. I'm going to rewrite this without EntityManager.

kaosko replied:
In fact, @Post.. transactions run in separate, cascading transaction. The previous one is already committed, then a new one is started. It'd be interesting to track down what exactly causes the ConcurrentModificationException. Yes, I'm well aware that this is not supported by standard JPA.

dmitrygusev commented on src/test/java/net/satago/tapestry5/jpa/test/JpaTest.java:
```diff
@@ -0,0 +1,187 @@
+		CriteriaBuilder qb = em.getCriteriaBuilder();
+		CriteriaQuery<T> query = qb.createQuery(type);
+		return em.createQuery(query).getResultList();
+	}
```
this method fails on line `em.createQuery(query).getResultList()` with:

> java.lang.IllegalStateException: No criteria query roots were specified

I had to rewrite it like this:

``` java
    private <T> List<T> getInstances(final Class<T> type)
    {
        EntityManager em = getEntityManager();

        if (em.getTransaction().isActive())
        {
            //  There shouldn't be any active transactions at this point,
            //  otherwise below query might return uncommitted changes
            em.getTransaction().rollback();
            em.clear();
        }

        CriteriaBuilder qb = em.getCriteriaBuilder();
        CriteriaQuery<T> query = qb.createQuery(type);
        Root<T> root = query.from(type);
        query.select(root);

        return em.createQuery(query).getResultList();
    }
```

http://stackoverflow.com/a/11670446

Note that transaction rollback is necessary if current transaction is active, otherwise assertions may be incorrect.

kaosko replied:
Regarding rollback, technically so but this is just scaffolding for the tests and there's nothing going on in parallel. Interesting that the implied root doesn't work on Hibernate - it's explicitly mentioned in Hibernate 3.5 docs (https://docs.jboss.org/hibernate/entitymanager/3.5/reference/en/html/querycriteria.html). Perhaps not supported yet by the version of Hibernate you are using.

dmitrygusev commented on src/test/java/net/satago/tapestry5/jpa/test/TopLevelServiceImpl.java:
```diff
@@ -0,0 +1,95 @@
+                // nestedService.createThingTwo(nameTwo);
+                ThingTwo thingTwo = new ThingTwo();
+                thingTwo.setName(nameTwo);
+                em.persist(thingTwo);
```
This test is not accurate, it's not clear what should happen here because there's no active transaction.

The `afterCommit` callback executed... after commit, outside of transaction scope, no new transaction will be open. And in JPA there should always be an open transaction, otherwise the behavior is not defined.

http://stackoverflow.com/a/21673304

dmitrygusev replied:
I believe we should uncomment `nestedService.createThingTwo(nameTwo);` and comment out the rest to fix this.

kaosko replied:
I agree. I was dealing with something else there and left it uncommented. However, your observation reveals a flaw in the logic - post commit invokables should not be automatically added to the sequential transaction deque.

dmitrygusev commented:
Should have referenced this PR instead of #2 from commit 950fbe26df5ff7dba1a79957ca7c9b3a7ef773c0.

I've back ported the tests to master (hence merge conflicts).

Mentioned commit should address most of above issues except when multiple different persistence contexts used in the same transactional stack -- this won't be supported unless there's a real need (i.e., all our applications use one persistence unit per application).

I'm thinking to add some sort of "strict mode" that would throw a `RuntimeException` if different PU names were used in the same stack. That mode should probably be enabled by default, and may be disabled with a symbol, or via `TransactionalUnitBuilder`'s API if user wants to force this for some reason. But this is a low priority now.

dmitrygusev commented:
Closing this PR without merge, because it has too many incompatible changes.

Most of the issues mentioned in this PR were addressed in commit 950fbe26df5ff7dba1a79957ca7c9b3a7ef773c0 (see previous comment) and covered with tests from this PR.

Work made in this PR was merged to the Tapestry5 repo via this commit https://github.com/apache/tapestry-5/commit/5047220cfe9a6f8a640fb10f0e3de266af51630f.

Thanks for working on this Kalle!

kaosko commented:
Agreed. Well thank you @dmitrygusev, while the implementation that ended up in the mainline differed a bit, the feature might never have found its way to Tapestry without you.
